### PR TITLE
Implementation for Amazon DynamoDB

### PR DESCRIPTION
**Pull request by nicktacular**
Added an implementation for Amazon DynamoDB. Using the `$options` key to the constructor, you can set which `$storageName` uses which key name for access. For example, if your key-val store called `Awesome` uses a key called `MyKey` then you can set it like so:

```php
$credentials = new \Aws\Credentials\Credentials('yo', 'sup');

$aws = new \Aws\Sdk([
    'timeout' => 1,
    'credentials' => $credentials,
    'region' => 'us-west-2',
    'version' => 'latest',
    'endpoint' => 'http://127.0.0.1:11888',
]);

$storage = new AmazonDynamoDbStorage(
    $aws->createDynamoDb(),
    null,
    ['storage_keys' => ['Awesome' => 'MyKey']]
);
```

Otherwise, the option `default_key_name` which defaults to `Id` will be used as the key name.

### CODE REVIEW

**Comment by nicktacular** [comment]
Oh, also fixes #41.

**Comment by beberlei** [comment]
@nicktacular is the storage thing sometihng we could better configure in the mappings, i.e. annotations xml and such?

Otherwise this looks really great, what do you think @EmanueleMinotto ?

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+     */
+    private $marshaler;
+
+    /**
+     * @var array
+     */
+    private $options = [
+        'default_key_name' => 'Id',
+        'storage_keys' => [],
+    ];
```
Comment (EmanueleMinotto) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Probably would be better to declare and define the properties `private $defaultKeyName = 'Id'` and `private $storageKeys = []`, for some reasons:
1. `$options` is used only by `AmazonDynamoDbStorage` so its content don't need to be flexible
2. if the final developer defines the *default_key_name* key as a non-scalar type, the class will generates different unexpected errors
3. a more strict type checking could be used asking for an `array $storageKeys = []`

What do you think about this @nicktacular?

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    /**
+     * Return a name of the underlying storage.
+     *
+     * @return string
+     */
+    public function getName()
+    {
+        return 'amazondynamodb';
```
Comment (EmanueleMinotto) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: I would use `amazon_dynamodb` here, to reproduce a sort of *[vendor]*_*[library]* like in the other implementations.

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+use Aws\DynamoDb\DynamoDbClient;
+use Aws\Result;
+use Doctrine\KeyValueStore\Storage\AmazonDynamoDbStorage;
+
+class AmazonDynamoDbTest extends \PHPUnit_Framework_TestCase
+{
+    private function getDynamoDbMock($methods = [])
+    {
+        $client = $this->getMockBuilder(DynamoDbClient::class)->disableOriginalConstructor();
```
Comment (EmanueleMinotto) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: Use `'Aws\DynamoDb\DynamoDbClient'` instead of `DynamoDbClient::class` please, so we can keep the php 5.4 support also on dev. :)

The same for every other `::class` used below.

**Comment by EmanueleMinotto** [comment]
@nicktacular I'm not sure if I'm understanding the part:
> Using the $options key to the constructor, you can set which $storageName uses which key name for access.

In this moment, looking the implementation, it seems a lot like the [`DoctrineCacheStorage::flattenKey`](https://github.com/doctrine/KeyValueStore/blob/master/lib/Doctrine/KeyValueStore/Storage/DoctrineCacheStorage.php#L62-L74), can you provide a more concrete example please?
It could be useful for documentation too.

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php (hunk 1 above)
Comment (nicktacular) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Agreed, will implement strict checking as well as separate these into different arrays.

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php (hunk 2 above)
Comment (nicktacular) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Agreed, will fix this.

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php (hunk 3 above)
Comment (nicktacular) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: Ah yes, will replace with classname as a string instead.

**Comment by nicktacular** [comment]
@EmanueleMinotto Sorry, I wasn't very clear there. What I meant to say is that the default name of the key is configurable. Also, if you want to override that key for any particular table (i.e. `$storageName`) then you can do that as well. Once I fix the format of the constructor and the way this is stored, I think it'll be clearer.

**Comment by nicktacular** [comment]
I believe I've addressed all the concerns except for this comment from @beberlei:

> is the storage thing sometihng we could better configure in the mappings, i.e. annotations xml and such?

I agree that perhaps this would be a good place handle this. Can you give me an example of what you'd like the annotation to look like?

Also, I'm wondering if I should write an integration test for an end-to-end test, configurable via PHPUnit with `<var name="DOCTRINE_KEYVALUE_AMAZON_DYNAMODB_ENDPOINT" value="http://localhost:8000" />` (among other properties) where, if you wish, I can write this test to work against some endpoint. One example could be to use [DynamoDB local](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tools.DynamoDBLocal.html) for this and skip those tests when this isn't configured. Thoughts on this?

**Comment by beberlei** [comment]
@nicktacular Integration test sounds great that only runs when all required environment variables are set, use env variables for this please, because other tools can do that more easily.

**Comment by EmanueleMinotto** [comment]
@nicktacular what's the status of the e2e test addition?

File: composer.json [hunk 4]
```diff
@@ -2,7 +2,8 @@
     "name": "doctrine/key-value-store",
     "require": {
         "php": ">=5.5",
-        "doctrine/common": "^2.4"
+        "doctrine/common": "^2.4",
+        "aws/aws-sdk-php": "^3.8"
```
Comment (Ocramius) [inline composer.json]: must be `require-dev`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        array $tableKeys = []
+    ) {
+        $this->client = $client;
+        $this->marshaler = $marshaler ?: new Marshaler();
+
+        if ($defaultKeyName !== null) {
+            $this->setDefaultKeyName($defaultKeyName);
+        }
+
+        if (!empty($tableKeys)) {
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: No need for this check

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    /**
+     * Validates a DynamoDB key name.
+     *
+     * @param $name mixed The name to validate.
+     *
+     * @throws \InvalidArgumentException When the key name is invalid.
+     */
+    protected function validateKeyName($name)
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: `private`

**Comment by nicktacular** [comment]
Sorry for the delay. Working on some issues with work; will be back to completing these tests shortly.

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    {
+        if (!is_string($name)) {
+            throw new \InvalidArgumentException(
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Please throw specialized exceptions, not SPL exceptions (write a new exception class - you may look at http://rosstuck.com/formatting-exception-messages/ )

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    /**
+     * Validates a DynamoDB table name.
+     *
+     * @see http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html
+     *
+     * @param $name string The table name to validate.
+     *
+     * @throws \InvalidArgumentException When the name is invalid.
+     */
+    protected function validateTableName($name)
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: `private`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    /**
+     * Sets the default key name for storage tables.
+     *
+     * @param $name string The default name to use for the key.
+     *
+     * @throws \InvalidArgumentException When the key name is invalid.
+     */
+    public function setDefaultKeyName($name)
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: `private`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+        $this->validateKeyName($name);
+        $this->defaultKeyName = $name;
+    }
+
+    /**
+     * Retrieves the default key name.
+     *
+     * @return string The default key name.
+     */
+    public function getDefaultKeyName()
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Needed?

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+
+    /**
+     * Sets a key name for a specific table.
+     *
+     * @param $table string The name of the table.
+     * @param $key string The name of the string.
+     *
+     * @throws \InvalidArgumentException When the key or table name is invalid.
+     */
+    public function setKeyForTable($table, $key)
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: `private`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+    /**
+     * Retrieves a specific name for a key for a given table. The default is returned if this table does not have
+     * an actual override.
+     *
+     * @param string $tableName The name of the table.
+     *
+     * @return string
+     */
+    public function getKeyNameForTable($tableName)
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: needed?

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+
+    /**
+     * Prepares a key to be in a valid format for lookups for DynamoDB. If passing an array, that means that the key
+     * is the name of the key and the value is the actual value for the lookup.
+     *
+     * @param string $storageName
+     *
+     * @return array The key in DynamoDB format.
+     */
+    protected function prepareKey($storageName, $key)
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: `private`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    /**
+     * Determine if the storage supports updating only a subset of properties,
+     * or if all properties have to be set, even if only a subset of properties
+     * changed.
+     *
+     * @return bool
+     */
+    public function supportsPartialUpdates()
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: add an `{@inheritDoc}`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+    {
+        //This is not true, but partial updates are too complicated given the available interface,
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: This comment should be at method level

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+        } else {
+            $keyValue = $key;
+            $keyName = $this->getKeyNameForTable($storageName);
+        }
+
+        return $this->marshaler->marshalItem([$keyName => $keyValue]);
+    }
+
+    /**
+     * Determine if the storage supports updating only a subset of properties,
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Don't duplicate the parent docblock

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+     */
+    public function supportsPartialUpdates()
+    {
+        //This is not true, but partial updates are too complicated given the available interface,
+        //meaning, the abstraction is insufficiently flexible enough to support this type of action.
+        return false;
+    }
+
+    /**
+     * Does this storage support composite primary keys?
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Don't duplicate the parent docblock, use `{@inheritDoc}`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 18]
```diff
@@ -0,0 +1,311 @@
...
+     *
+     * @return bool
+     */
+    public function supportsCompositePrimaryKeys()
+    {
+        return true;
+    }
+
+    /**
+     * Does this storage require composite primary keys?
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Don't duplicate the parent docblock, use `{@inheritDoc}`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+     * Insert data into the storage key specified.
+     *
+     * @param string       $storageName
+     * @param array|string $key
+     * @param array        $data
+     */
+    public function insert($storageName, $key, array $data)
+    {
+        $this->client->putItem([
+            'TableName' => $storageName,
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Consider using class constants for these - makes it easier to refactor and to understand/document the meaning

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+     *
+     * @return bool
+     */
+    public function requiresCompositePrimaryKeys()
+    {
+        return false;
+    }
+
+    /**
+     * Insert data into the storage key specified.
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Don't duplicate the parent docblock, use `{@inheritDoc}`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+    public function insert($storageName, $key, array $data)
+    {
+        $this->client->putItem([
+            'TableName' => $storageName,
+            'Item' => $this->prepareKey($storageName, $key) + $this->marshaler->marshalItem($data),
+        ]);
+    }
+
+    /**
+     * Update data into the given key.
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Don't duplicate the parent docblock, use `{@inheritDoc}`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+     * @param array        $data
+     */
+    public function update($storageName, $key, array $data)
+    {
+        //We are using PUT so we just replace the original item
+        $this->insert($storageName, $key, $data);
+    }
+
+    /**
+     * Delete data at key.
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Don't duplicate the parent docblock, use `{@inheritDoc}`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+     * Update data into the given key.
+     *
+     * @param string       $storageName
+     * @param array|string $key
+     * @param array        $data
+     */
+    public function update($storageName, $key, array $data)
+    {
+        //We are using PUT so we just replace the original item
+        $this->insert($storageName, $key, $data);
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: What if the key doesn't exist?

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+    public function delete($storageName, $key)
+    {
+        $this->client->deleteItem([
+            'TableName' => $storageName,
+            'Key' => $this->prepareKey($storageName, $key),
+        ]);
+    }
+
+    /**
+     * Find data at key.
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Don't duplicate the parent docblock, use `{@inheritDoc}`

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 25]
```diff
@@ -0,0 +1,311 @@
...
+            'Key' => $this->prepareKey($storageName, $key),
+        ]);
+
+        if (!$item) {
+            throw new NotFoundException();
+        }
+
+        $item = $item->get('Item');
+        if (!is_array($item)) {
+            throw new NotFoundException();
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Maybe more customized exception message in this case?

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 26]
```diff
@@ -0,0 +1,311 @@
...
+    public function find($storageName, $key)
+    {
+        $item = $this->client->getItem([
+            'TableName' => $storageName,
+            'ConsistentRead' => true,
+            'Key' => $this->prepareKey($storageName, $key),
+        ]);
+
+        if (!$item) {
+            throw new NotFoundException();
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Maybe include the key in the exception? Otherwise, hard to debug

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 27]
```diff
@@ -0,0 +1,311 @@
...
+        $item = $item->get('Item');
+        if (!is_array($item)) {
+            throw new NotFoundException();
+        }
+
+        return $this->marshaler->unmarshalItem($item);
+    }
+
+    /**
+     * Return a name of the underlying storage.
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Don't duplicate the parent docblock, use `{@inheritDoc}`

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php [hunk 28]
```diff
@@ -0,0 +1,278 @@
...
+     * @expectedExceptionMessage The key must be a string, got "array" instead.
+     */
+    public function testDefaultKeyCannotBeSomethingOtherThanString()
+    {
+        $client = $this->getDynamoDbMock();
+        new AmazonDynamoDbStorage($client, null, []);
+    }
+
+    /**
+     * @expectedException \InvalidArgumentException
```
Comment (Ocramius) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: Please use `$this->setExpectedException()` instead, so you know exactly at which point in the execution an exception should be thrown

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php [hunk 29]
```diff
@@ -0,0 +1,278 @@
+<?php
+
+namespace Doctrine\Tests\KeyValueStore\Storage;
+
+use Doctrine\KeyValueStore\Storage\AmazonDynamoDbStorage;
+
+class AmazonDynamoDbTest extends \PHPUnit_Framework_TestCase
```
Comment (Ocramius) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: Add `@covers \Doctrine\KeyValueStore\Storage\AmazonDynamoDbStorage` to avoid coverage leakage

**Comment by nicktacular** [comment]
Thanks for the feedback @Ocramius; I'll implement.

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php [hunk 30]
```diff
@@ -0,0 +1,278 @@
+<?php
+
```
Comment (Ocramius) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: License headers

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php [hunk 31]
```diff
@@ -0,0 +1,278 @@
...
+     * @expectedExceptionMessage The key must be a string, got "object" instead.
+     */
+    public function testTableKeysMustAllBeStringsOrElse()
+    {
+        $client = $this->getDynamoDbMock();
+        new AmazonDynamoDbStorage($client, null, null, ['mytable' => 'hello', 'yourtable' => new \stdClass()]);
+    }
+
+    /**
+     * @expectedException \InvalidArgumentException
```
Comment (Ocramius) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: same as above

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php [hunk 32]
```diff
@@ -0,0 +1,278 @@
...
+        return [
+            ['MyTable'],
+            ['This_is0k-...'],
+            ['hello_world'],
+            ['...........00....']
+        ];
+    }
+
+    /**
+     * @dataProvider invalidTableNames
```
Comment (Ocramius) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: same as above

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php [hunk 33]
```diff
@@ -0,0 +1,278 @@
...
+     * @expectedException \InvalidArgumentException
```
Comment (Ocramius) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: same as above

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php [hunk 34]
```diff
@@ -0,0 +1,278 @@
...
+            'TableName' => 'MyTable',
+            'Key' => ['Id' => ['S' => 'abc123']]
+        ]));
+
+        $storage = new AmazonDynamoDbStorage($client);
+        $storage->delete('MyTable', 'abc123');
+    }
+
+    /**
+     * @expectedException \Doctrine\KeyValueStore\NotFoundException
```
Comment (Ocramius) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: same as above

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 35]
```diff
@@ -0,0 +1,311 @@
...
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\KeyValueStore\Storage;
+
+use Aws\DynamoDb\DynamoDbClient;
+use Aws\DynamoDb\Marshaler;
+use Doctrine\KeyValueStore\NotFoundException;
+
+class AmazonDynamoDbStorage implements Storage
```
Comment (Ocramius) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: can be `final` if all the API is defined in `Storage`

**Comment by Ocramius** [comment]
@nicktacular thanks for the patience, and sorry for the noise :-) Just ordinary administration here

**Comment by nicktacular** [comment]
@EmanueleMinotto I didn't realize there was a deadline. What's the time you need this done by?

**Comment by EmanueleMinotto** [comment]
@nicktacular no deadline, I'm just trying to reach a RC before the end of the year :)
https://github.com/doctrine/KeyValueStore/milestones

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php (hunk 23 above)
Comment (nicktacular) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: If an item by that key does not exist, it is simply inserted. Is that your question?

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php (hunk 25 above)
Comment (nicktacular) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Actually, I don't even know if this is possible. I can't find whether or not this type of response is possible from AWS for DynamoDb, so I'm inclined to just remove this check. Thoughts?

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php (hunk 35 above)
Comment (nicktacular) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: Not sure I quite follow. By definition, since this is not abstract, the API must be fully honored, but I don't think it's a good idea to make an implementation final if someone wants to extend that behavior. Perhaps I'm not understanding your request?

File: tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php (hunk 30 above)
Comment (EmanueleMinotto) [inline tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php]: @nicktacular after a rebase you can just run `php-cs-fixer fix`

**Comment by nicktacular** [comment]
@EmanueleMinotto - the eventual consistency model is proving to be a significant hurdle in the functional tests. I will continue working on this, but wanted to let you know.

**Comment by EmanueleMinotto** [comment]
@nicktacular don't worry, that's not a problem, thank you for your effort (ping me on IRC if you need a help) :)

**Comment by EmanueleMinotto** [comment]
@nicktacular do you have news about this? If you can't, than I could merge this into a separate branch and complete it by myself :)

**Comment by nicktacular** [comment]
@EmanueleMinotto I've been really sick recently. I also couldn't get integration tests to work, so I'm going to implement using DynamoDb local store instead. I will let you know as soon as I'm done.

**Comment by EmanueleMinotto** [comment]
oh ok, no problem
thanks @nicktacular, let me know :)

**Comment by nicktacular** [comment]
Rewrote functional tests to utilize DynamoDb local (removed waits and such), but now it seems to be randomly failing, possibly to do with something on OS X as per [this AWS thread](https://forums.aws.amazon.com/thread.jspa?threadID=157820). I'm going to bring up an Ubuntu machine and try there.

**Comment by nicktacular** [comment]
Ok so it's not Dynamo. It's `BasicCrudTestCase` class which seems to be taking an inordinate amount on time on something. It's approximately 17 seconds per test and it seems to be hung up on the call:

```php
$this->manager = $this->createManager($this->storage, $mappingDriver);
```

For what reason I can't understand, but I'm going to dig into this now to see why this is happening. Updates to follow...

**Comment by nicktacular** [comment]
Well, that was quick. When commenting out the `xml` mapping driver, the test speeds up immensely. So, I'm going to hazard a guess that the XSD reference in the `Post` fixture for XML mapping is what's causing this hangup:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://doctrine-project.org/schemas/orm/doctrine-mapping
                          http://raw.github.com/doctrine/doctrine2/master/doctrine-mapping.xsd">
	<entity name="Doctrine\Tests\KeyValueStore\Functional\Post" storage-name="post">
		<id>id</id>
	</entity>
</doctrine-mapping>
```

**Comment by nicktacular** [comment]
Ok, so I've discovered a conceptual issue with tables in DynamoDb. How do we want to deal with mismatches on types for ids? Suppose, for example, the type is a string id in DynamoDb and the id is set to an integer. Should the type of the primary key be deduced by means of a `describeTable` operation on the table first, or provide it as a configuration option?

When this occurs, the raw exception from DynamoDb looks like this:

```
Client error: `POST http://localhost:8000` resulted in a `400 Bad Request` response:
{"__type":"com.amazon.coral.validate#ValidationException","message":"Type mismatch for attribute to update"}
```

Doing a `describeTable` operation seems wasteful, so I'd prefer to rely on configuration (i.e. annotation) instead. Any thoughts on this?

**Comment by nicktacular** [comment]
Having thought about this more, I think that the right approach is to use an appropriate exception message to indicate that there is an attribute mismatch and leave it up to the consumer for resolving this problem. Additionally, a new configuration will need to be set to indicate the type of key that is on the table (or keys).

Furthermore, an additional test will need to be added to test composite keys which is the error @tiagobrito mentions. **However**, it's only feasible to deal with this error when the table and entity actually have composite keys. If there is a mismatch, it will still be up to the developer to address this issue and ensure that the table key schema and entity match.

Finally, I'm not convinced that these configuration options make sense to be annotations. These configurations are too specific to the adapter, so I think they should be adapter configurations and not moved to annotations.

In summary:

 * Will add configuration options so that the developer can configure the key or keys and the type for each key (string vs numeric) and the associated tests.
 * Will add functional tests for composite keys.
 * Will add functional tests for mismatches which will *necessarily* throw an exception with an adequate message explaining why this is an error and what is needed to address.

I will wait a few days before I continue this implementation in case there are any serious concerns with this approach.

**Comment by EmanueleMinotto** [comment]
well done @nicktacular, I'm also against the usage of a new annotation for a single driver, so :+1: for:

> deduced by means of a describeTable operation on the table first

or other configurations you think are required for the DynamoDB storage

File: lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php [hunk 36]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function insert($storageName, $key, array $data)
+    {
+        $this->client->putItem([
+            self::TABLE_NAME_KEY => $storageName,
+            self::TABLE_ITEM_KEY => $this->prepareKey($storageName, $key) + $this->marshaler->marshalItem($data),
```
Comment (larowlan) [inline lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php]: DynamoDb chokes on empty values.
E.g 
```php
$data = [
  'foo' => '',
  'bar' => 'baz',
];
```
Would choke. The [existing implementation](https://github.com/doctrine/KeyValueStore/blob/master/lib/Doctrine/KeyValueStore/Storage/DynamoDbStorage.php#L194) includes a `prepareData` method - however it only consider one-dimensional arrays, multi-dimensional cleaning is needed too.

e.g
```php
$data = [
  'foo' => [
     'baz' => '',
     'bar' => 'boo',
  ],
  'bar' => 'baz',
];
```
needs to be sent as
```php
$data = [
  'foo' => [
     'bar' => 'boo',
  ],
  'bar' => 'baz',
];
```

**Comment by larowlan** [comment]
Promised work is in https://github.com/nicktacular/KeyValueStore/pull/2, sent against @nicktacular branch so once combined will be available here

**Comment by larowlan** [comment]
Let me know if you'd prefer to close this and open a new one instead of sending fixes against @nicktacular branch
